tool/tsh/common/tsh.go

```diff
@@ -1471,44 +1439,92 @@ func Run(ctx context.Context, args []string, opts ...CliOption) error {
+	// in order to be able to access the exit code and stdout/stderr of the command
+	// that was run and determine if we should create a new access request from
+	// the output data.
+	// We don't want to enable tracing for the master invocation of tsh kubectl
+	// because the data that we would be tracing would be the tsh kubectl command.
+	// Instead, we want to enable tracing for the re-executed kubectl command and
+	// we do that in the kubectl command handler.
+	case cf.command == "kubectl":
+		return func() {}
+	// The user explicitly asked for traces to be sent to a particular exporter
+	// instead of forwarding them to Auth. Proceed with creating the provider.
+	case cf.SampleTraces && cf.TraceExporter != "":
 		provider, err := tracing.NewTraceProvider(cf.Context, tracing.Config{
-			Service:     teleport.ComponentTSH,
-			ExporterURL: cf.TraceExporter,
-			// We are using 1 here to record all spans as a result of this tsh command. Teleport
-			// will respect the recording flag of remote spans even if the spans it generates
-			// wouldn't otherwise be recorded due to its configured sampling rate.
-			SamplingRate: 1.0,
+			Service:      teleport.ComponentTSH,
+			ExporterURL:  cf.TraceExporter,
+			SamplingRate: samplingRate,
 		})
 
+		if err != nil {
+			log.WithError(err).Debugf("failed to connect to trace exporter %s", cf.TraceExporter)
+			return func() {}
+		}
+
 		cf.TracingProvider = provider
 		cf.tracer = provider.Tracer(teleport.ComponentTSH)
-		return provider, trace.Wrap(err)
-	}
-
-	// create a TeleportClient and generate a provider that forwards
-	// spans to Auth
+		return flush(provider)
+	// The login command cannot forward spans to Auth since there is no way to get
+	// an authenticated client to forward with until after the authentication ceremony
+	// is complete. However, if the user explicitly provided an exporter then the login
+	// spans can be sent directly to it.
+	case cf.command == "login":
+		return func() {}
+	// All commands besides ssh are only traced if the user explicitly requested
+	// tracing. For ssh, a random number of spans may be sampled if the Proxy is
+	// for a Cloud tenant.
+	case !cf.SampleTraces && !slices.Contains(autoForwardedToCloud, cf.command):
+		return func() {}
+	case cf.SampleTraces:
+	}
+
+	// Parse the config to determine if forwarding is needed for Cloud and
+	// to get a handle to an Auth client.
 	tc, err := makeClient(cf)
 	if err != nil {
-		return nil, trace.Wrap(err)
+		log.WithError(err).Debug("failed to set up span forwarding.")
+		return func() {}
+	}
+
+	if !cf.SampleTraces {
+		// Automatically enable and forward spans to Cloud if the following conditions are met:
+		// 1) The proxy address resembles that of a Cloud tenant
+		// 2) The command being executed is tsh ssh
+		// 3) The user has not already explicitly provided the --trace flag.
+		if strings.HasSuffix(tc.WebProxyAddr, ".teleport.sh:443") {
```

Comment:
Maybe this should be a dedicated function, and we can decide if we want to add staging/dev in there as well?
```suggestion
		if isCloudWebProxyAddr(tc.WebProxyAddr) {
```